### PR TITLE
Change publishers to publish unique ptrs

### PR DESCRIPTION
**Pull request by Sarath18**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #1672 |
| Primary OS tested on | Ubuntu 18.04 |
| Robotic platform tested on | gazebo simulation of TurtleBot3 (Waffle) |

---

## Description of contribution in a few bullet points
- Change all the publishers to publish unique pointers.

## Description of documentation updates required from your changes
- Publishers in navigation2 should publish unique pointers rather than raw messages.

---

## Future work that may be required in bullet points

- I have tried to cover all ROS publishers in navigation2. I'll make the necessary changes if any required.

### CODE REVIEW

File: nav2_system_tests/src/localization/test_localization_node.cpp [hunk 1]
```diff
@@ -53,7 +54,7 @@ class TestAmclPose : public ::testing::Test
     subscription_ = node->create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
       "amcl_pose", rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
       std::bind(&TestAmclPose::amcl_pose_callback, this, _1));
-    initial_pose_pub_->publish(testPose_);
+    initial_pose_pub_->publish(*testPose_);
```
Comment (fmrico) [inline nav2_system_tests/src/localization/test_localization_node.cpp:57]: isnt't it `initial_pose_pub_->publish(std::move(*testPose_))`?

File: nav2_system_tests/src/localization/test_localization_node.cpp [hunk 2]
```diff
@@ -81,15 +82,15 @@ class TestAmclPose : public ::testing::Test
 
 bool TestAmclPose::defaultAmclTest()
 {
-  initial_pose_pub_->publish(testPose_);
+  initial_pose_pub_->publish(*testPose_);
```
Comment (fmrico) [inline nav2_system_tests/src/localization/test_localization_node.cpp:85]: same here `initial_pose_pub_->publish(std::move(*testPose_))`?

File: nav2_system_tests/src/localization/test_localization_node.cpp [hunk 3]
```diff
@@ -81,15 +82,15 @@ class TestAmclPose : public ::testing::Test
...
   while (!pose_callback_) {
     // TODO(mhpanah): Initial pose should only be published once.
-    initial_pose_pub_->publish(testPose_);
+    initial_pose_pub_->publish(*testPose_);
```
Comment (fmrico) [inline nav2_system_tests/src/localization/test_localization_node.cpp:88]: same here `initial_pose_pub_->publish(std::move(*testPose_))`?

**Comment by Sarath18** [comment]
I realized a few mistakes on my side. Firstly, I shouldn't have modified the code and the test cases together.

Secondly, modifications to this [test case](https://github.com/ros-planning/navigation2/blob/5c175587a76608932d274444b716fa3e8161faa6/nav2_system_tests/src/localization/test_localization_node.cpp) essentially do the same thing. During publishing, it is dereferencing the pointer which leads to the publishing of raw message instead. 

> isnt't it initial_pose_pub_->publish(std::move(*testPose_))?

@fmrico, let's look at the changes you suggested together. `testPose_` is a unique_ptr inside the class and modifications to `testPose_` happens in multiple functions of the class. So when the publish statement is called, the pointer gets dereferenced and then the data is moved to the publisher. Now the block that contained the data is deallocated and the pointer is still valid. This resulted in a failure of the test case. Even `initial_pose_pub_->publish(std::move(testPose_));` won't work as this statement will pass the ownership to the publisher and `testPose_` will become a null pointer.

I am unclear on how to tackle this problem. If `testPose_` was a normal data member instead of a pointer, how do publish a unique pointer when publish statement is called?
The [current implementation](https://github.com/ros-planning/navigation2/blob/5c175587a76608932d274444b716fa3e8161faa6/nav2_costmap_2d/src/costmap_2d_publisher.cpp#L174) was to create a unique_ptr using the raw message and then publish it. But this might problems again as it will return a pointer to a newly allocated memory after moving the contents from the variable. This will deallocate that data member. Looking at it again I found out that this implementation worked out because data is being populated every time before being published.
Please do correct me if I am wrong.

**Comment by fmrico** [comment]
Hi @Sarath18,

You are right!! But in this case, I suggest not to change this file. As you said, this is test code and maybe it shouldn't be changed.

I have run your code in sim and seems to be working properly. IMHO you did great work!! 

I have done a little experiment comparing your improved code and the baseline. The robot performs a navigation along three waypoint, as shown in the desktop capture. The files contain a record every second of cpu (2nd column) and mem (3rd column).

Hope it helps!!
[cpu_baseline.txt](https://github.com/ros-planning/navigation2/files/4569994/cpu_baseline.txt)
[cpu_improve.txt](https://github.com/ros-planning/navigation2/files/4569995/cpu_improve.txt)
Captura de pantalla de 2020-05-03 12-14-28

**Comment by Sarath18** [comment]
> You are right!! But in this case, I suggest not to change this file. As you said, this is test code and maybe it shouldn't be changed.

I should revert back the test cases.

> I have run your code in sim and seems to be working properly. IMHO you did great work!!

Great! Thanks :)

> I have done a little experiment comparing your improved code and the baseline. The robot performs a navigation along three waypoint, as shown in the desktop capture. The files contain a record every second of cpu (2nd column) and mem (3rd column).

Thank you for benchmarking the updates. Can you please clarify what baseline is? Is without the updates? Looking at the results you provided, I noticed that in general `cpu_improve.txt` shows more cpu usage compared to the baseline while memory usage remains the same. Weird.

**Comment by fmrico** [comment]
Hi @Sarath18 

Maybe we should repeat this experiment. I think that the first values of improved (this PR) are affected in some way by the execution of the experiment with baseline (master). 

Anyway, If you see the sequence discarding first results, using unique_ptr significative reduces the cpu computation (second column). There are no significant changes in the mem computation (3rd column).

If you want to obtain your own values, in the screenshot you can find the command that I used to obtain these results.

**Comment by Sarath18** [comment]
Hi @fmrico,
> Anyway, If you see the sequence discarding first results, using unique_ptr significative reduces the cpu computation (second column). There are no significant changes in the mem computation (3rd column).

Thank you for clarifying. I made a plot of your results and that helped me understand the CPU utilization difference.
plot_legend

> The current implementation was to create a unique_ptr using the raw message and then publish it. But this might problems again as it will return a pointer to a newly allocated memory after moving the contents from the variable. This will deallocate that data member. Looking at it again I found out that this implementation worked out because data is being populated every time before being published.

Any ideas on how to solve `Costmap` and `Grid` publishers in `costmap_2d_publisher.cpp`? @SteveMacenski  mentioned in the ticket:
> We should transition all ROS publishers in navigation (and anywhere we can) to this method. This is **especially important for topics publishing larger messages like pointclouds, maps, and images.**

File: nav2_recoveries/plugins/back_up.cpp [hunk 4]
```diff
@@ -77,23 +78,23 @@ Status BackUp::onCycleUpdate()
...
+  cmd_vel->angular.z = 0.0;
+  command_x_ < 0 ? cmd_vel->linear.x = -command_speed_ : cmd_vel->linear.x = command_speed_;
 
   geometry_msgs::msg::Pose2D pose2d;
   pose2d.x = current_pose.pose.position.x;
   pose2d.y = current_pose.pose.position.y;
   pose2d.theta = tf2::getYaw(current_pose.pose.orientation);
 
-  if (!isCollisionFree(distance, cmd_vel, pose2d)) {
+  if (!isCollisionFree(distance, *cmd_vel, pose2d)) {
```
Comment (SteveMacenski) [inline nav2_recoveries/plugins/back_up.cpp:91]: You can change the signature of this function to use the unique_ptr

File: nav2_recoveries/plugins/spin.cpp [hunk 5]
```diff
@@ -106,21 +107,21 @@ Status Spin::onCycleUpdate()
...
+  auto cmd_vel = std::make_unique<geometry_msgs::msg::Twist>();
+  cmd_yaw_ < 0 ? cmd_vel->angular.z = -vel : cmd_vel->angular.z = vel;
 
   geometry_msgs::msg::Pose2D pose2d;
   pose2d.x = current_pose.pose.position.x;
   pose2d.y = current_pose.pose.position.y;
   pose2d.theta = tf2::getYaw(current_pose.pose.orientation);
 
-  if (!isCollisionFree(relative_yaw, cmd_vel, pose2d)) {
+  if (!isCollisionFree(relative_yaw, *cmd_vel, pose2d)) {
```
Comment (SteveMacenski) [inline nav2_recoveries/plugins/spin.cpp:118]: As well here, please change the function signature

File: nav2_system_tests/src/localization/test_localization_node.cpp (hunk 1 above)
Comment (SteveMacenski) [reply nav2_system_tests/src/localization/test_localization_node.cpp]: Not quite, you're moving the pointer ownership, not the object.

File: nav2_system_tests/src/localization/test_localization_node.cpp (hunk 2 above)
Comment (SteveMacenski) [reply nav2_system_tests/src/localization/test_localization_node.cpp]: Not quite, you're moving the pointer ownership, not the object.

File: nav2_system_tests/src/localization/test_localization_node.cpp (hunk 3 above)
Comment (SteveMacenski) [reply nav2_system_tests/src/localization/test_localization_node.cpp]: Not quite, you're moving the pointer ownership, not the object.

**Comment by SteveMacenski** [comment]
@Sarath18 when you do std::move it nulls out the pointer, so it needs to be reallocated. For all the cases when you publish something, you need to make sure that the variable is local to the function so that a new unique_ptr will be made each time its called or that the class variables are reallocated before a next call. 

i'm not overly concerned with changing the tests because those aren't used in production and this is a production optimization.

File: nav2_costmap_2d/plugins/voxel_layer.cpp [hunk 6]
```diff
@@ -400,7 +401,8 @@ void VoxelLayer::raytraceFreespace(
     clearing_endpoints_.header.frame_id = global_frame_;
     clearing_endpoints_.header.stamp = clearing_observation.cloud_->header.stamp;
 
-    clearing_endpoints_pub_->publish(clearing_endpoints_);
+    auto cloud_msg = std::make_unique<sensor_msgs::msg::PointCloud>(std::move(clearing_endpoints_));
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/voxel_layer.cpp:404]: Why is this move required?

File: nav2_costmap_2d/src/costmap_2d_publisher.cpp [hunk 7]
```diff
@@ -170,7 +170,9 @@ void Costmap2DPublisher::publishCostmap()
 {
   if (node_->count_subscribers(costmap_raw_pub_->get_topic_name()) > 0) {
     prepareCostmap();
-    costmap_raw_pub_->publish(costmap_raw_);
+
+    auto costmap_raw_ptr = std::make_unique<nav2_msgs::msg::Costmap>(std::move(costmap_raw_));
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_2d_publisher.cpp:174]: This seems unnecessary. What's the type of costmap_raw_? Can you make it into a unique pointer to begin with in the class variable? Then you just create the unique_ptr in the prepareCostmap and then just std::move to publish it. Right now, you're making 2 copies.

File: nav2_costmap_2d/plugins/voxel_layer.cpp (hunk 6 above)
Comment (Sarath18) [reply nav2_costmap_2d/plugins/voxel_layer.cpp:404]: Actually, `clearing_endpoints_` is not being used anywhere outside the function `VoxelLayer::raytraceFreespace`. I can try converting it to a unique_ptr local to this function.

File: nav2_costmap_2d/src/costmap_2d_publisher.cpp (hunk 7 above)
Comment (Sarath18) [reply nav2_costmap_2d/src/costmap_2d_publisher.cpp:174]: Yes, that is better. I should make the similar changes with `prepareGrid()` function as well.

**Comment by Sarath18** [comment]
> For all the cases when you publish something, you need to make sure that the variable is local to the function

Got it.

> i'm not overly concerned with changing the tests because those aren't used in production and this is a production optimization.

I reverted back the test case anyway.

> Those charts do look suspect. I don't see why gazebo would drop so much CPU based on a far downstream change. I think this would have to be re-run at least 5 times on the same computer to get results.

I'll rerun the tests as you said on my computer and then share the results. @fmrico, it would be great if you could throw some more light on this.

> @fmrico what script did you use for that? Can you share it with us? 

You can find the script in the screenshot @fmrico attached [above](https://user-images.githubusercontent.com/3810011/80911625-c1a8bc00-8d37-11ea-86cb-52d0abc58ad6.png). 
```bash
while true; do ps -a -o comm= -o pcpu= -o pmem=; echo "================================"; sleep 1; done >> cpu_baseline.txt
```
<br>

> I would be really interested in this analysis though (and a number like saved N%), that could make a very useful write up in the ROS2 documentation and if we had tangible proof of its impact, that could be really valuable.

> Might be useful to add to the tools/ directory. If we can get the recording code and then a plotting code, that could be useful for making optimizations.

I believe we both have written rough scripts to test this feature. If you really want to add it to the tools directory we can polish the logging/visualization before that. I think this can be moved to a separate ticket to populate ideas and incorporate more features.

**Comment by fmrico** [comment]
Hi @SteveMacenski 

As @Sarath18 wrote, the command used was very simple: 

`while true; do ps -a -o comm= -o pcpu= -o pmem=; echo "================================"; sleep 1; done >> cpu_baseline.txt
`

It's simple, but it gives us a quick idea o the improvement

Francisco

**Comment by Sarath18** [comment]
> I see 2 publishes missing in the map server for the map and in the planner server publishing paths.

Hey @SteveMacenski, I didn't update these publishers to publish unique_ptrs because they were not in the local scope of the publisher and were being used in multiple places. I see a few ways to update this but I am not sure if that is the best way to do so. `result` in planner_server is a shared pointer being passed to action_server_ and `msg_` (occupancy_grid) in map_server is a unique pointer that is being updated in different methods of the class.

I have made all other requested changes.

**Comment by SteveMacenski** [comment]
I think the map server `msg_` is only published by 1 entity (correct me if I'm wrong). You should be able to create the unique_ptr on bringup, and then after `std::move` to publish, instantiate it again. If everything is single threaded, then there's no risk of someone calling on that nullptr. 

For the planner the _path_ isn't a shared pointer, the result is a shared pointer. You can change the path return from the planner to be a unique_ptr local variable that's dereferenced and copied to the result shared pointer. There's going to be 1 copy somewhere, lets just own it in the code explicitly.

File: nav2_costmap_2d/src/costmap_2d_publisher.cpp [hunk 8]
```diff
@@ -107,85 +107,93 @@ void Costmap2DPublisher::onNewSubscription(const ros::SingleSubscriberPublisher&
...
-void Costmap2DPublisher::prepareGrid()
+std::unique_ptr<nav_msgs::msg::OccupancyGrid> Costmap2DPublisher::prepareGrid()
 {
   std::unique_lock<Costmap2D::mutex_t> lock(*(costmap_->getMutex()));
-  double resolution = costmap_->getResolution();
+  grid_resolution = costmap_->getResolution();
+  grid_width = costmap_->getSizeInCellsX();
+  grid_height = costmap_->getSizeInCellsY();
+
+  auto grid_ = std::make_unique<nav_msgs::msg::OccupancyGrid>();
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_2d_publisher.cpp:117]: You can leave grid_ as a class variable, in fact, I'd prefer it. You just need to `std::make_unique<nav_msgs::msg::OccupancyGrid>();` on the class variable at the start of prepare grid to instantiate it. I'd revert this commit and just make the simpler change here and in publish grid. You don't want to return smart pointer types.

**Comment by SteveMacenski** [comment]
Looks like you have some merge conflicts (somehow?) with amcl

**Comment by Sarath18** [comment]
> For the planner the path isn't a shared pointer, the result is a shared pointer. You can change the path return from the planner to be a unique_ptr local variable that's dereferenced and copied to the result shared pointer. There's going to be 1 copy somewhere, lets just own it in the code explicitly.

If there is going to be one copy somewhere, why not do it just before the publish (just before or inside `publishPlan`)? By doing this we can incorporate any updates to the path (if any) before publishing. I tested it out and it works just fine.

> I think the map server msg_ is only published by 1 entity (correct me if I'm wrong).

There are two publishers in the MapServerClass. One int the `on_activate` lifecycle state and the other one in the `loadMapCallback`. 

> You should be able to create the unique_ptr on bringup, and then after std::move to publish, instantiate it again. If everything is single threaded, then there's no risk of someone calling on that nullptr.

I did this. But I had to populate the OccupancyGrid by calling `loadMapFromYaml` after instantiation. I don't think this is a good way to do that. By having another class variable that stores the OccupancyGrid and then copying it into unique_ptr we will again face a 1 copy somewhere.

> Looks like you have some merge conflicts (somehow?) with amcl

This is because of the update to amcl particle cloud to incorporate weights. I'll merge this once everything else is solved. 

I have made all other requested changes.

**Comment by SteveMacenski** [comment]
> If there is going to be one copy somewhere, why not do it just before the publish (just before or inside publishPlan)? By doing this we can incorporate any updates to the path (if any) before publishing. I tested it out and it works just fine.

You could also do `make_unique<...>(result->path)` as well, sure. I was just pointing out that the path wasn't the shared ptr, it was a member of one so there's no a conversion from shared_ptr -> unique_ptr required. Its just copying the contents of a memory of the result to the unique_ptr to publish.

> There are two publishers in the MapServerClass. One int the on_activate lifecycle state and the other one in the loadMapCallback.

~That's fine. In both places do~ 

```
occ_pub_->publish(std::move(msg_));
msg_ = ... make_unique<>();
```

~The only issue you would run into is if the activate was called at the exact same time as `loadMapCallback`, which I don't think is possible since the on_activate is called from a callback that will be taking up the node executor so no other service calls can be made, and vise versa.~

> But I had to populate the OccupancyGrid by calling loadMapFromYaml after instantiation.

You are correct. Mhm. That makes map server really interesting. Maybe we should change the map `msg_` to be a stack variable and we copy it to a unique_ptr to publish.

**Comment by Sarath18** [comment]
@SteveMacenski, I have made all the changes. If everything seems fine, I'll merge master here and fix any conflicts that occur.

I am not sure if I encounter a bug or not but `ament_cpplint` was not configured for `nav2_costmap_2d` when running tests. I had to manually run the checks and resolve the issues in that package.

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_publisher.hpp [hunk 9]
```diff
@@ -151,8 +151,10 @@ class Costmap2DPublisher
   // Service for getting the costmaps
   rclcpp::Service<nav2_msgs::srv::GetCostmap>::SharedPtr costmap_service_;
 
-  nav_msgs::msg::OccupancyGrid grid_;
-  nav2_msgs::msg::Costmap costmap_raw_;
+  float grid_resolution;
```
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_publisher.hpp:154]: I'm not entirely clear why these were added. Can you explain these new member variables? Looking at the use, it seems like you can just use the information in their scopes.

**Comment by SteveMacenski** [comment]
> I am not sure if I encounter a bug or not but ament_cpplint was not configured for nav2_costmap_2d when running tests. I had to manually run the checks and resolve the issues in that package.

You're welcome to add that then.

Also, you still have conflicts to resolve.

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_publisher.hpp (hunk 9 above)
Comment (Sarath18) [reply nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_publisher.hpp:154]: To store the previous state of grid i.e. resolution, width and height. 
If you look at [this](https://github.com/ros-planning/navigation2/blob/628bf78e75e733433c6500aa8ccea2574d65eb16/nav2_costmap_2d/src/costmap_2d_publisher.cpp#L184-L193) part of the code, grid resolution, width and height are being checked for in the if condition which is before the `prepareGrid()` call, where `grid_` is actually instantiated. The std::move in the publish call nulls out the `grid_` pointer, so I don't have the required information when this condition is checked the next time.

Comment (SteveMacenski) [reply nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_publisher.hpp:154]: Ok thats fine

Comment (SteveMacenski) [reply nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_publisher.hpp:154]: Just need merge conflicts and linters enabled and good to go

**Comment by AlexeyMerzlyakov** [comment]
> @AlexeyMerzlyakov can you review the map server part?

Checked. Map server change looks good to me. I agree that there is no need to keep `msg_` to be unique_ptr since it will be converted to rvalue later.

**Comment by Sarath18** [comment]
Thank you for this opportunity. This helped me to understand the navigation2 structure and the inner working much more clearly. Along with this, I got to learn about smart pointers, memory management, code practices and conventions too :smile:.
> That makes map server really interesting. 

I agree, but the entire navigation stack is one beautiful piece of art and it just keeps getting better :star2:

**Pull request merged**